### PR TITLE
Upgrade to syn 2.0

### PR DESCRIPTION
**Pull request by lnicola**

### CODE REVIEW

**Comment by lnicola** [comment]
This compiles, but doesn't work because of a run-time breaking change in `syn`:

```
error: expected parentheses
 --> tests/macros.rs:3:5
  |
3 |     #[salsa::invoke(another_module::another_name)]
  |     ^
```

**Comment by lnicola** [comment]
CC @fee1-dead maybe you ran into this before?

File: components/salsa-macros/src/query_group.rs [hunk 1]
```diff
@@ -681,13 +681,13 @@ impl TryFrom<syn::Attribute> for SalsaAttr {
...
-        if is_not_salsa_attr_path(&attr.path) {
+        if is_not_salsa_attr_path(&attr.path()) {
             return Err(attr);
         }
 
         let span = attr.span();
-        let name = attr.path.segments[1].ident.to_string();
-        let tts = attr.tokens.into();
+        let name = attr.path().segments[1].ident.to_string();
+        let tts = attr.into_token_stream().into();
```
Comment (fee1-dead) [inline components/salsa-macros/src/query_group.rs:690]: Doesn't this return the entire attribute? Previously `tokens` were the tokens that are not the path e.g. `(foo)` in `salsa::something(foo)`.

Comment (lnicola) [reply components/salsa-macros/src/query_group.rs:690]: Yeah, that's probably it. I looked at `attr.meta`, but couldn't really make it work.

**Comment by Veykril** [comment]
Would be nice if we could land https://github.com/salsa-rs/salsa/pull/443 here as well should we cut another pre-release for old salsa (I don't think syn 2 demands changes for that PR)

File: components/salsa-macros/src/query_group.rs (hunk 1 above)
Comment (fee1-dead) [reply components/salsa-macros/src/query_group.rs]: The `MetaList` works differently. It doesn't have the path, but it doesn't have the parenthesis either. So you would have `foo`. You could probably `quote!` this and wrap the result around parentheses as a dirty hack.

Comment (lnicola) [reply components/salsa-macros/src/query_group.rs]: This is actually working now AFAICT, it fails to build because of some `Span`/`DelimSpan` changes.

**Comment by fee1-dead** [comment]
I think you could just use the `DelimSpan`'s two spans in an array `[span.open(), span.close()]` and that should work

**Comment by lnicola** [comment]
Thanks, that leaves [this bit](https://github.com/salsa-rs/salsa/blob/master/components/salsa-2022-macros/src/jar.rs#L161-L163) here, where `Paren::span` is a `DelimSpan`, but `Ident::span()` returns a `Span` and the only way to create a `DelimSpan` is from a `Group`.

**Comment by fee1-dead** [comment]
Try using the [`Group`](https://github.com/dtolnay/syn/blob/1316ec6d1787d78d9e5722d23205f1c9395e068f/src/token.rs#L617-L623) function.

**Comment by fee1-dead** [comment]
actually wait nvm. you wanted to create a paren token

**Comment by fee1-dead** [comment]
Ah, I think you can should still be able to call `token::Paren(span)` with a single span because of the impl [here](https://github.com/dtolnay/syn/blob/1316ec6d1787d78d9e5722d23205f1c9395e068f/src/span.rs#L50-L56).

**Comment by lnicola** [comment]
Thanks, this was such a struggle :pensive:.

**Comment by nikomatsakis** [comment]
Hi @lnicola -- I apologize for letting this sit for so long. If you are still interested, would you consider rebasing it? If not, I understand.

**Comment by fee1-dead** [comment]
I cloned the repository, wanting to rebase this PR, but saw that the macros already use syn 2.0. Therefore there's no need to rebase this now.

**Comment by lnicola** [comment]
Yup, I guess this is no longer necessary?

**Pull request closed without merging**